rfc/0011-audit-checks.md

```diff
@@ -0,0 +1,354 @@
+While the Kubernetes API server uses `/validate/POL_ID`, the audit scanner
+will file its requests against `/audit/POL_ID` endpoint.
+
+The `/audit` endpoint will be different from the `/validate` one in these regards:
+
+* Generate different Prometheus metrics: we don't want the overall
+  `/validate` metrics to be polluted by the audit checks
+* Generate different Open Telemetry traces: again, these should not be mixed
+  with the `/validate` ones. We might even want to turn them off to make the
+  audit execution faster
+* Policies in `monitor` mode should still behave like the ones in `protect`
+  mode. They have to produce rejections if needed
+
+By issuing the audit request against a real Policy Server, we solve these problems:
+
+* Policies can be hosted on registries secured with credentials. We don't have
+  to worry about being able to fetch the WebAssembly modules from these
+  authenticated registries
+* Policies might be hosted on insecure registires/registires with custom TLS
+  certificates. Again, we won't have to bother about that
+* Some data used by the policy (like external OCI manifests, sigstore
+  verification results) can be cached by the Policy Server and reused between
+  the audit evaluations and the regular validations requests
+
+Sending all our auditing requests to the actual Policy Server could lead to poor
+performance of the regular admission requests.
+Currently, based on our load testing data, one single Policy Server running with
+2 workers can evaluate about 1600 requests per second.
+Based on this data, we think the requests generated by the audit checks should
+not clog the Policy Server.
+If that turns out to be true, we can embrace different strategies to mitigate
+that.
+
+Until we have real world data, it's not useful to eagerly optimize this aspect.
+
+### Step 6: write `PolicyReport`
+
+The code will overwrite the old `PolicyReport` instance with a new generated
+from the data kept in memory. By doing that, we don't have to worry about
+synchronizing the old contents with the new ones.
+
+## Changes to Policy Server
+
+A new rest API endpoint is going to be created called `/audit`.
+
+The endpoint will receive POST requests similar to the ones generated by the
+Kubernetes API server.
+
+All the policies hosted by the Policy Server will be exposed via `/audit/POL_ID`
+endpoints.
+
+The `/audit` API will behave in a slightly different way compared to the current `/validate`
+endpoint. The differences have been illustrated by the previous section of things
+RFC.
+
+## Changes to ClusterAdmissionPolicy and AdmissionPolicy CRDs
+
+It's necessary a change in the policy CRDs to mark a policy to be run in the background audit checks.
+The field would be called `backgroundAudit` and its default value will be `true`.
+Therefore, all the policies will run in the audit checks by default, unless the user
+configure the other way. Example of policies CRDs with the `backgroundAudit` field:
```

Comment:
Are we gonna add the `backgroundAudit` field to metadata.yaml as we do with `mutating`?

Reply:
I would add it. This is now mentioned in a new section I've just added